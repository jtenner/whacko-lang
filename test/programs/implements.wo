export interface MyInterface {
  field: i32;
  method(a: i64): i32;
}

export class MyClass implements MyInterface {
  field: i32;
  constructor() {
    this.field = i32(42);
  }
  method(a: i64): i32 {
    this.field = i32(a);
    return i32(2);
  }
}

export class MyOtherClass implements MyInterface {
  field: i32;
  constructor() {
    this.field = i32(42);
  }
  method(a: i64): i32 {
    this.field = i32(a);
    return i32(1);
  }
}

export fn useMyInterface(a: MyInterface): i32 {
  a.field = i32(0);
  return a.method(69);
}

export fn _start(): void {
  let a = new MyClass();
  let b = new MyOtherClass();
  useMyInterface(a);
}
