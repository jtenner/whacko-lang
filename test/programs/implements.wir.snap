program:
  callable:[std/builtins.wo]gc.initialize;functionType:()=>voidType
(extern function)
  callable:[std/builtins.wo]gc.alloc;functionType:(u32Type,usizeType)=>ptrType
(extern function)
  callable:[std/builtins.wo]gc.store_field_ptr;functionType:(i32Type,ptrType,ptrType,ptrType)=>voidType
(extern function)
  callable:[test/programs/implements.wo]_start;functionType:()=>voidType
    block: entry~329
      335~inst = new: classType:[test/programs/implements.wo]MyClass<>
      336~inst = call: callable:[test/programs/implements.wo]MyClass.constructor;methodType:classType:[test/programs/implements.wo]MyClass<>#()=>voidType with (runtime classType:[test/programs/implements.wo]MyClass<> 335~inst)
      337~inst = store: (variable a) = (runtime classType:[test/programs/implements.wo]MyClass<> 335~inst)
      341~inst = new: classType:[test/programs/implements.wo]MyOtherClass<>
      342~inst = call: callable:[test/programs/implements.wo]MyOtherClass.constructor;methodType:classType:[test/programs/implements.wo]MyOtherClass<>#()=>voidType with (runtime classType:[test/programs/implements.wo]MyOtherClass<> 341~inst)
      343~inst = store: (variable b) = (runtime classType:[test/programs/implements.wo]MyOtherClass<> 341~inst)
      344~inst = load: (variable a)
      348~inst = call: callable:[test/programs/implements.wo]useMyInterface;functionType:(interfaceType:[test/programs/implements.wo]MyInterface<>)=>i32Type with (runtime classType:[test/programs/implements.wo]MyClass<> 344~inst)
      return (void)

  callable:[test/programs/implements.wo]MyClass.constructor;methodType:classType:[test/programs/implements.wo]MyClass<>#()=>voidType
    block: entry~378
      379~inst = load: (variable const compiler-generated)
      380~inst = load: (field classType:[test/programs/implements.wo]MyClass<>#field of (runtime classType:[test/programs/implements.wo]MyClass<> 379~inst))
      385~inst = (i32 42)
      386~inst = store: (field classType:[test/programs/implements.wo]MyClass<>#field of (runtime classType:[test/programs/implements.wo]MyClass<> 379~inst)) = (runtime i32Type 385~inst)
      return (void)

  callable:[test/programs/implements.wo]MyOtherClass.constructor;methodType:classType:[test/programs/implements.wo]MyOtherClass<>#()=>voidType
    block: entry~368
      369~inst = load: (variable const compiler-generated)
      370~inst = load: (field classType:[test/programs/implements.wo]MyOtherClass<>#field of (runtime classType:[test/programs/implements.wo]MyOtherClass<> 369~inst))
      375~inst = (i32 42)
      376~inst = store: (field classType:[test/programs/implements.wo]MyOtherClass<>#field of (runtime classType:[test/programs/implements.wo]MyOtherClass<> 369~inst)) = (runtime i32Type 375~inst)
      return (void)

  callable:[test/programs/implements.wo]useMyInterface;functionType:(interfaceType:[test/programs/implements.wo]MyInterface<>)=>i32Type
    block: entry~350
      351~inst = load: (variable a)
      352~inst = load: (field interfaceType:[test/programs/implements.wo]MyInterface<>#field of (runtime interfaceType:[test/programs/implements.wo]MyInterface<> 351~inst))
      357~inst = (i32 0)
      358~inst = store: (field interfaceType:[test/programs/implements.wo]MyInterface<>#field of (runtime interfaceType:[test/programs/implements.wo]MyInterface<> 351~inst)) = (runtime i32Type 357~inst)
      359~inst = load: (variable a)
      365~inst = (i64 69)
      366~inst = call: callable:[test/programs/implements.wo]MyInterface.method;methodType:interfaceType:[test/programs/implements.wo]MyInterface<>#(i64Type)=>i32Type with (runtime interfaceType:[test/programs/implements.wo]MyInterface<> 359~inst), (runtime i64Type 365~inst)
      return (runtime i32Type 366~inst)

  callable:[test/programs/implements.wo]MyInterface.method;methodType:interfaceType:[test/programs/implements.wo]MyInterface<>#(i64Type)=>i32Type
(interface method)
  callable:[test/programs/implements.wo]MyClass.method;methodType:classType:[test/programs/implements.wo]MyClass<>#(i64Type)=>i32Type
    block: entry~411
      413~inst = load: (variable const compiler-generated)
      414~inst = load: (field classType:[test/programs/implements.wo]MyClass<>#field of (runtime classType:[test/programs/implements.wo]MyClass<> 413~inst))
      415~inst = load: (variable a)
      419~inst = cast: <i32Type> [object Object]
      420~inst = store: (field classType:[test/programs/implements.wo]MyClass<>#field of (runtime classType:[test/programs/implements.wo]MyClass<> 413~inst)) = (runtime i32Type 419~inst)
      return (i32 2)

  callable:[test/programs/implements.wo]MyOtherClass.method;methodType:classType:[test/programs/implements.wo]MyOtherClass<>#(i64Type)=>i32Type
    block: entry~396
      398~inst = load: (variable const compiler-generated)
      399~inst = load: (field classType:[test/programs/implements.wo]MyOtherClass<>#field of (runtime classType:[test/programs/implements.wo]MyOtherClass<> 398~inst))
      400~inst = load: (variable a)
      404~inst = cast: <i32Type> [object Object]
      405~inst = store: (field classType:[test/programs/implements.wo]MyOtherClass<>#field of (runtime classType:[test/programs/implements.wo]MyOtherClass<> 398~inst)) = (runtime i32Type 404~inst)
      return (i32 1)

